File: android/app/src/main/kotlin/app/hapi/companion/feature/chat/blocks/ToolContentPresentation.kt
package app.hapi.companion.feature.chat.blocks

import app.hapi.protocol.chat.ChatToolCall
import app.hapi.protocol.chat.getInputStringAny
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/** Presentation aliases only; never rewrite the recorded call or grouping input. */
internal fun toolPresentationName(name: String): String = when (val local = name.removePrefix("functions.")) {
    "exec_command" -> "Bash"
    "apply_patch" -> "CodexPatch"
    else -> local
}

internal val TERMINAL_TOOLS = setOf("Bash", "CodexBash", "shell_command", "run_shell_command", "write_stdin")

internal fun toolSourceInput(input: JsonElement?, keys: List<String>): String? =
    input.textValue() ?: getInputStringAny(input, keys)

private fun JsonElement?.textValue(): String? = (this as? JsonPrimitive)?.takeIf { it.isString }?.content

internal fun isPlanProposalTool(name: String): Boolean = name == "ExitPlanMode" || name == "exit_plan_mode"

/** A proposal is input.plan Markdown, not update_plan's checklist or a result. */
internal fun planProposalMarkdown(tool: ChatToolCall): String? =
    if (isPlanProposalTool(tool.name)) (tool.input as? JsonObject)?.get("plan").textValue()?.takeIf { it.isNotBlank() }
    else null

internal fun planProposalShowsResult(tool: ChatToolCall): Boolean {
    if (tool.state == "error") return true
    val result = tool.result ?: return false
    if (result is JsonNull) return false
    return result.textValue()?.isNotBlank() ?: true
}

/** Mixed content stays JSON, rather than silently discarding images/resources. */
internal fun extractResultText(result: JsonElement, depth: Int = 0): String? {
    if (depth > 4) return null
    return when (result) {
        is JsonPrimitive -> result.textValue()
        is JsonArray -> {
            val texts = result.map { entry ->
                entry.textValue() ?: run {
                    val obj = entry as? JsonObject ?: return null
                    if (obj["type"] != null && obj["type"].textValue() != "text") return null
                    obj["text"].textValue() ?: return null
                }
            }
            texts.joinToString("\n")
        }
        is JsonObject -> {
            val stdout = result["stdout"].textValue()
            val stderr = result["stderr"].textValue()
            if (stdout != null || stderr != null) {
                val parts = mutableListOf<String>()
                stdout?.takeIf { it.isNotEmpty() }?.let(parts::add)
                stderr?.takeIf { it.isNotEmpty() }?.let { parts.add("stderr:\n$it") }
                if (parts.isNotEmpty()) return parts.joinToString("\n\n")
            }
            (result["file"] as? JsonObject)?.get("content").textValue()?.let { return it }
            var emptyText: String? = null
            for (key in listOf("content", "text", "output", "error", "message", "result", "data")) {
                val value = result[key] ?: continue
                if (value is JsonNull) continue
                val text = extractResultText(value, depth + 1)
                if (!text.isNullOrEmpty()) return text
                if (text != null) emptyText = ""
                if (key == "content" && value is JsonArray && text == null) return null
            }
            if (stdout != null || stderr != null) "" else emptyText
        }
    }
}

internal sealed interface ToolResultStyle {
    data object Terminal : ToolResultStyle
    data class Code(val language: String?) : ToolResultStyle
    data object Markdown : ToolResultStyle
}

internal fun toolResultStyle(tool: ChatToolCall): ToolResultStyle {
    val name = toolPresentationName(tool.name)
    if (tool.state == "error") return ToolResultStyle.Terminal
    val parsed = ((tool.input as? JsonObject)?.get("parsed_cmd") as? JsonArray)?.singleOrNull()
    val readCommand = name == "CodexBash" && getInputStringAny(parsed, listOf("type")) == "read"
    if (name in setOf("Read", "NotebookRead") || readCommand) {
        val file = (tool.result as? JsonObject)?.get("file")
        val path = getInputStringAny(file, listOf("filePath", "file_path"))
            ?: getInputStringAny(tool.input, listOf("file_path", "path", "file", "notebook_path"))
            ?: if (readCommand) getInputStringAny(parsed, listOf("name", "path", "file_path")) else null
        return ToolResultStyle.Code(languageForPath(path))
    }
    if (name in setOf("WebFetch", "WebSearch", "Task", "Agent", "Skill", "ExitPlanMode", "exit_plan_mode")) {
        return ToolResultStyle.Markdown
    }
    return ToolResultStyle.Terminal
}

internal const val TOOL_TEXT_PAGE_SIZE = 20_000

/** No truncation; page boundaries never split UTF-16 surrogate pairs. */
internal fun toolTextPages(text: String): List<String> = buildList {
    var start = 0
    while (start < text.length) {
        var end = minOf(start + TOOL_TEXT_PAGE_SIZE, text.length)
        if (end < text.length && Character.isHighSurrogate(text[end - 1]) && Character.isLowSurrogate(text[end])) end--
        add(text.substring(start, end))
        start = end
    }
}

internal fun toolQuestionOptionText(option: JsonElement): String {
    option.textValue()?.let { return it }
    val label = getInputStringAny(option, listOf("label", "value")) ?: return prettyJson(option)
    val description = getInputStringAny(option, listOf("description"))
    return if (description.isNullOrEmpty()) label else "$label — $description"
}

/** Exit/status information remains visible even when stdout is empty. */
internal fun toolResultMetadata(result: JsonElement?, depth: Int = 0): List<String> {
    if (depth > 4) return emptyList()
    val obj = result as? JsonObject ?: return emptyList()
    val fields = listOf("exit_code" to listOf("exit_code", "exitCode"), "status" to listOf("status"),
        "session_id" to listOf("session_id"), "wall_time_seconds" to listOf("wall_time_seconds"))
    val metadata = fields.mapNotNull { (label, keys) ->
        val value = keys.firstNotNullOfOrNull { key ->
            (obj[key] as? JsonPrimitive)?.takeIf { it !is JsonNull &&
                (it.isString && it.content.isNotEmpty() || !it.isString && it.content.toDoubleOrNull() != null) }
        }
        value?.let { "$label: ${it.content}" }
    }
    if (metadata.isNotEmpty()) return metadata
    for (key in listOf("output", "result", "data")) {
        val nested = toolResultMetadata(obj[key], depth + 1)
        if (nested.isNotEmpty()) return nested
    }
    return emptyList()
}

File: ios/Hapi/Features/Chat/Blocks/ToolContentPresentation.swift
import Foundation
import HapiProtocol

/// Presentation aliases only. Keep the recorded name/input and chat grouping intact.
func toolPresentationName(_ name: String) -> String {
    let name = name.hasPrefix("functions.") ? String(name.dropFirst("functions.".count)) : name
    switch name {
    case "exec_command": return "Bash"
    case "apply_patch": return "CodexPatch"
    default: return name
    }
}

let terminalToolNames: Set<String> = [
    "Bash", "CodexBash", "shell_command", "run_shell_command", "write_stdin",
]

func toolSourceInput(_ input: JSONValue?, keys: [String]) -> String? {
    input?.chatString ?? chatInputString(input, keys)
}

func isPlanProposalTool(_ name: String) -> Bool {
    name == "ExitPlanMode" || name == "exit_plan_mode"
}

/// Plans are documents in input.plan, not update_plan's checklist or a tool result.
func planProposalMarkdown(_ tool: ChatToolCall) -> String? {
    guard isPlanProposalTool(tool.name),
          let plan = tool.input?[chatKey: "plan"]?.chatString,
          !plan.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return nil }
    return plan
}

/// A synthesized Codex proposal completes with null output. Keep diagnostics,
/// but don't suggest that its visible document is still waiting for output.
func planProposalShowsResult(_ tool: ChatToolCall) -> Bool {
    if tool.state == .error { return true }
    guard let result = tool.result, result != .null else { return false }
    if let text = result.chatString {
        return !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }
    return true
}

/// Bounded wrapper traversal; mixed/non-text blocks deliberately retain their
/// JSON representation rather than silently dropping images or resources.
func extractResultText(_ result: JSONValue, depth: Int = 0) -> String? {
    guard depth <= 4 else { return nil }
    switch result {
    case .string(let text): return text
    case .array(let entries):
        var texts: [String] = []
        for entry in entries {
            if let text = entry.chatString {
                texts.append(text)
            } else if let object = entry.chatObject,
                      object["type"] == nil || object["type"]?.chatString == "text",
                      let text = object["text"]?.chatString {
                texts.append(text)
            } else {
                return nil
            }
        }
        return texts.joined(separator: "\n")
    case .object(let object):
        let stdout = object["stdout"]?.chatString
        let stderr = object["stderr"]?.chatString
        if stdout != nil || stderr != nil {
            var parts: [String] = []
            if let out = stdout, !out.isEmpty { parts.append(out) }
            if let err = stderr, !err.isEmpty { parts.append("stderr:\n\(err)") }
            if !parts.isEmpty { return parts.joined(separator: "\n\n") }
            // Empty stdout must not hide a structured error.
        }
        if let content = object["file"]?[chatKey: "content"]?.chatString { return content }
        var emptyText: String?
        for key in ["content", "text", "output", "error", "message", "result", "data"] {
            if let value = object[key], value != .null {
                let text = extractResultText(value, depth: depth + 1)
                if let text, !text.isEmpty { return text }
                if text != nil { emptyText = "" }
                // Preserve a mixed content array in full, not just another field.
                if key == "content", case .array = value, text == nil {
                    return nil
                }
            }
        }
        if stdout != nil || stderr != nil { return "" }
        return emptyText
    default: return nil
    }
}

enum ToolResultStyle: Equatable {
    case terminal
    case code(String?)
    case markdown
}

func toolResultStyle(_ tool: ChatToolCall) -> ToolResultStyle {
    let name = toolPresentationName(tool.name)
    // Errors are literal diagnostics, not markdown or source code.
    if tool.state == .error { return .terminal }
    let parsed = tool.input?[chatKey: "parsed_cmd"]?.chatArray
    let readCommand = name == "CodexBash" && parsed?.count == 1 && parsed?.first?[chatKey: "type"]?.chatString == "read"
    if ["Read", "NotebookRead"].contains(name) || readCommand {
        let file = tool.result?[chatKey: "file"]
        let path = chatInputString(file, ["filePath", "file_path"])
            ?? chatInputString(tool.input, ["file_path", "path", "file", "notebook_path"])
            ?? (readCommand ? chatInputString(parsed?.first, ["name", "path", "file_path"]) : nil)
        return .code(languageForPath(path))
    }
    if ["WebFetch", "WebSearch", "Task", "Agent", "Skill", "ExitPlanMode", "exit_plan_mode"].contains(name) {
        return .markdown
    }
    return .terminal
}

func toolQuestionOptionText(_ option: JSONValue) -> String {
    if let text = option.chatString { return text }
    if let label = chatInputString(option, ["label", "value"]) {
        if let description = chatInputString(option, ["description"]), !description.isEmpty {
            return "\(label) — \(description)"
        }
        return label
    }
    return chatPrettyJSON(option)
}

/// Keep command status visible even when stdout is empty. Technical wire labels
/// avoid conflating the process exit code with the tool call's transport state.
func toolResultMetadata(_ result: JSONValue?, depth: Int = 0) -> [String] {
    guard depth <= 4, let object = result?.chatObject else { return [] }
    let fields = [("exit_code", ["exit_code", "exitCode"]), ("status", ["status"]),
                  ("session_id", ["session_id"]), ("wall_time_seconds", ["wall_time_seconds"])]
    let metadata = fields.compactMap { label, keys -> String? in
        for key in keys {
            guard let value = object[key] else { continue }
            switch value {
            case .string(let text) where !text.isEmpty: return "\(label): \(text)"
            case .number: return "\(label): \(chatPrettyJSON(value))"
            default: continue
            }
        }
        return nil
    }
    if !metadata.isEmpty { return metadata }
    for key in ["output", "result", "data"] {
        let nested = toolResultMetadata(object[key], depth: depth + 1)
        if !nested.isEmpty { return nested }
    }
    return []
}
